### PR TITLE
Improve erc20 unlock

### PR DESCRIPTION
**Pull request by paouvrard**
- Build proofs after user calls `unlock` instead of during `checkSync`: fixes https://github.com/aurora-is-near/rainbow-bridge-client/issues/33
- Check sync with the withdraw receipt instead of with the current NEAR finality: fixes the problem where users had to wait for the next light client block before they could finalize a recovered transfer.
- Check if an unlock proof was already used:
    - Don't prompt user signature and mark transfer as complete if the proof was already used.
    - When recovering a transfer, check if it is complete.

Checking if proofs have already been used on the NEAR side may take longer to implement as the token factory should be upgraded with a view function. So we may want to merge this first.

### CODE REVIEW

File: packages/nep141-erc20/src/bridged-nep141/sendToEthereum/index.js [hunk 1]
```diff
@@ -533,49 +549,46 @@ async function checkSync (transfer) {
...
+ * @param {*} proof
+ */
+async function proofAlreadyUsed (web3, proof) {
+  const usedProofsKey = bs58.decode(proof.outcome_proof.outcome.receipt_ids[0]).toString('hex')
+  const usedProofsMappingPosition = '0'.repeat(63) + '3'
+  const storageIndex = web3.utils.sha3('0x' + usedProofsKey + usedProofsMappingPosition)
+  // eth_getStorageAt docs: https://eth.wiki/json-rpc/API
+  const proofIsUsed = await web3.eth.getStorageAt(process.env.ethLockerAddress, storageIndex)
+  return Number(proofIsUsed) === 1
+}
```
Comment (mfornet) [inline packages/nep141-erc20/src/bridged-nep141/sendToEthereum/index.js:590]: Didi you tested manually that this is working?

File: packages/nep141-erc20/src/bridged-nep141/sendToEthereum/index.js [hunk 2]
```diff
@@ -584,23 +597,35 @@ async function checkSync (transfer) {
...
 async function unlock (transfer) {
   const web3 = new Web3(getEthProvider())
-  const ethUserAddress = (await web3.eth.getAccounts())[0]
 
+  // Build burn proof
+  transfer = await checkSync(transfer)
+  const proof = await findProof(transfer)
+
+  if (await proofAlreadyUsed(web3, proof)) {
+    // TODO find the unlockTxHash (requires ERC20Locker upgrade)
```
Comment (mfornet) [inline packages/nep141-erc20/src/bridged-nep141/sendToEthereum/index.js:606]: Please open an issue in https://github.com/aurora-is-near/rainbow-token-connector and let's add a link here to that issue

**Review by mfornet — APPROVED** [review]

File: packages/nep141-erc20/src/bridged-nep141/sendToEthereum/index.js (hunk 1 above)
Comment (paouvrard) [reply packages/nep141-erc20/src/bridged-nep141/sendToEthereum/index.js:590]: Yes, tested with multiple transfers

File: packages/nep141-erc20/src/bridged-nep141/sendToEthereum/index.js (hunk 2 above)
Comment (paouvrard) [reply packages/nep141-erc20/src/bridged-nep141/sendToEthereum/index.js:606]: https://github.com/aurora-is-near/rainbow-token-connector/issues/59

**Pull request merged**